lib/krane/kubernetes_resource.rb

```diff
@@ -409,17 +413,44 @@ def self.go_template_for(kind, name)
       def self.extract_all_from_go_template_blob(blob)
         blob.split(EVENT_SEPARATOR).map do |event_blob|
           pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
+          count, timestamp = extract_count_timestamp(pieces)
+
           new(
             subject_kind: pieces[FIELDS.index(".involvedObject.kind")],
             subject_name: pieces[FIELDS.index(".involvedObject.name")],
-            count: pieces[FIELDS.index(".count")],
-            last_timestamp: pieces[FIELDS.index(".lastTimestamp")],
+            count: count,
+            last_timestamp: timestamp,
             reason: pieces[FIELDS.index(".reason")],
             message: pieces[FIELDS.index(".message")]
           )
         end
       end
 
+      def self.extract_count_timestamp(pieces)
+        series = pieces[FIELDS.index(".series")]
+        last_timestamp = pieces[FIELDS.index(".lastTimestamp")]
+        deprecated_timestamp = pieces[FIELDS.index(".deprecatedLastTimestamp")]
+
+        # Find the right count and last_timestamp according to Kubernetes API and kubectl version
+        if deprecated_timestamp.present? && deprecated_timestamp != "<no value>"
+          # kubectl < 1.16 uses events.k8s.io/v1beta1, which has .deprecatedCount and
+          # .deprecatedLastTimestamp fields
+          [pieces[FIELDS.index(".deprecatedCount")], deprecated_timestamp]
```

Comment:
I'm worried about coupling count to timestap. Given we've seen cases where they're both blank, it seems possible one could be but not the other one.  Can you also explain why you chose this ordering of cases?

Reply:
I broke the code to find the right event count and timestamp in two methods and I followed this order:

1) First deprecated fields `.deprecatedCount` and `.deprecatedLastTimestamp` because Kubernetes 1.16 is the only version that does not uses it, therefore, the deprecated fields are the ones that are more likely to appear.
2) If the above fields aren't present, I check for k8s 1.16 standard API fields
3) If the above fields aren't present, I check for the presence of .series field, which is present in k8s 1.1.6 as well
4) If none of those fields are present, I fallback to count as `1` and `.eventTime`, as in this case in Kubernetes `1.16`:

```
» kubectl get events -o json --namespace=k8sdeploy-test-pvc-no-pv-dafc5d11a073e407 --context=minikube --request-timeout=15
{
    "apiVersion": "v1",
    "items": [
        {
            "action": "Scheduling",
            "apiVersion": "v1",
            "eventTime": "2019-11-09T02:27:17.835909Z",
            "firstTimestamp": null,
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "Pod",
                "name": "pvc",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37187",
                "uid": "c07e276b-6ea9-4b27-801d-42949cfdc9c9"
            },
            "kind": "Event",
            "lastTimestamp": null,
            "message": "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:17Z",
                "name": "pvc.15d55d7d5c41277b",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37189",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/pvc.15d55d7d5c41277b",
                "uid": "1369fe84-67c1-4c43-826b-6ba2c33402ef"
            },
            "reason": "FailedScheduling",
            "reportingComponent": "default-scheduler",
            "reportingInstance": "default-scheduler-minikube",
            "source": {
                "component": "default-scheduler"
            },
            "type": "Warning"
        },
        {
            "action": "Scheduling",
            "apiVersion": "v1",
            "eventTime": "2019-11-09T02:27:17.851047Z",
            "firstTimestamp": null,
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "Pod",
                "name": "pvc",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37188",
                "uid": "c07e276b-6ea9-4b27-801d-42949cfdc9c9"
            },
            "kind": "Event",
            "lastTimestamp": null,
            "message": "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:17Z",
                "name": "pvc.15d55d7d5d2828fa",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37194",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/pvc.15d55d7d5d2828fa",
                "uid": "8dc9a49e-66b2-444c-9c31-217cd8268ec7"
            },
            "reason": "FailedScheduling",
            "reportingComponent": "default-scheduler",
            "reportingInstance": "default-scheduler-minikube",
            "series": {
                "count": 1,
                "lastObservedTime": "2019-11-09T02:27:19.264307Z"
            },
            "source": {
                "component": "default-scheduler"
            },
            "type": "Warning"
        },
        {
            "apiVersion": "v1",
            "count": 3,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "with-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37173",
                "uid": "00c29fd4-8104-400f-b41f-d22012b01bee"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:44Z",
            "message": "waiting for first consumer to be created before binding",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "with-storage-class.15d55d7d0620c54f",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37238",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/with-storage-class.15d55d7d0620c54f",
                "uid": "a06ce861-756a-4660-9a7f-db53f083db15"
            },
            "reason": "WaitForFirstConsumer",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "persistentvolume-controller"
            },
            "type": "Normal"
        },
        {
            "apiVersion": "v1",
            "count": 2,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "without-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37176",
                "uid": "d0d2ff61-52e1-4850-a632-687b0f89ec8f"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:16Z",
            "message": "waiting for a volume to be created, either by external provisioner \"k8s.io/minikube-hostpath\" or manually created by system administrator",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "without-storage-class.15d55d7d07392317",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37179",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/without-storage-class.15d55d7d07392317",
                "uid": "b5d75be0-c579-407d-bd53-e9828ea7d4de"
            },
            "reason": "ExternalProvisioning",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "persistentvolume-controller"
            },
            "type": "Normal"
        },
        {
            "apiVersion": "v1",
            "count": 1,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "without-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37175",
                "uid": "d0d2ff61-52e1-4850-a632-687b0f89ec8f"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:16Z",
            "message": "External provisioner is provisioning volume for claim \"k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/without-storage-class\"",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "without-storage-class.15d55d7d07f1f2b0",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37180",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/without-storage-class.15d55d7d07f1f2b0",
                "uid": "6b5c8e3f-a7e2-4278-94dc-8afae68a94a6"
            },
            "reason": "Provisioning",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "k8s.io/minikube-hostpath c9d9d89f-01d8-11ea-9057-aa43d77eee0d"
            },
            "type": "Normal"
        },
        {
            "apiVersion": "v1",
            "count": 1,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "without-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37175",
                "uid": "d0d2ff61-52e1-4850-a632-687b0f89ec8f"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:16Z",
            "message": "Successfully provisioned volume pvc-d0d2ff61-52e1-4850-a632-687b0f89ec8f",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "without-storage-class.15d55d7d08aef384",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37184",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/without-storage-class.15d55d7d08aef384",
                "uid": "ed4f557a-bd08-4726-8802-fdd173cc7a33"
            },
            "reason": "ProvisioningSucceeded",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "k8s.io/minikube-hostpath c9d9d89f-01d8-11ea-9057-aa43d77eee0d"
            },
            "type": "Normal"
        }
    ],
    "kind": "List",
    "metadata": {
        "resourceVersion": "",
        "selfLink": ""
    }
}
```

Reply:
I get what you're saying, but its confusing becasuse you don't seperate when you mean 
kubectl 1.16 preferred version vs resources version supported on k8s 1.16

Reply:
I reordered the extraction logic to follow what should correct k8s API order: currently supported fields, series for k8s 1.16, deprecated fields and fallback.